pandas/core/indexes/datetimelike.py

```diff
@@ -816,7 +809,10 @@ def _fast_union(self, other, sort=None):
             loc = right.searchsorted(left_end, side="right")
             right_chunk = right.values[loc:]
             dates = concat_compat((left.values, right_chunk))
-            return self._shallow_copy(dates)
+            result = self._shallow_copy(dates)
+            result._set_freq("infer")
```

Comment:
you could change _set_freq to return self to make this idiom more clear (also _set_freq should be marked in the doc-string as a mutating function)

Reply:
Holding off on this since _set_freq needs a bigger rethink, xref #31218.  

All other comments addressed, i think